### PR TITLE
Azure Provider: return to using token's groups claim

### LINKED ISSUE #1868: [doc] Azure AD permission incorrect?
## Expected Behavior
Setting the permission to fetch groups per the documentation, which should be `User.Read` for Delegated and `Directory.Read.All` for Application.

https://learn.microsoft.com/en-us/graph/api/user-list-transitivememberof?view=graph-rest-1.0&tabs=http#permissions

## Current Behavior

`Group.Read.All` is listed.

https://github.com/oauth2-proxy/oauth2-proxy/blob/aafa96655048bc1ed26487e92410f5fe64968fa2/docs/docs/configuration/auth.md?plain=1#L82

## Context
https://learn.microsoft.com/en-us/graph/api/user-list-transitivememberof endpoint is used to fetch group membership.

https://github.com/oauth2-proxy/oauth2-proxy/blob/aafa96655048bc1ed26487e92410f5fe64968fa2/providers/azure.go#L137

### PR DESCRIPTION
**Pull request by hskiba**
## Description

Return to using the groups claim in the Azure provider. This involved the following changes:

1. Switch to using the [Microsoft Graph SDK for Go](https://github.com/microsoftgraph/msgraph-sdk-go) for graph API queries.
2. Use the `groups` claim if it's present.
3. Only query the Microsoft Graph API in the following cases:
   1. `--azure-graph-group-field` is `displayName`. It's only possible to return group IDs in a groups claim (see https://github.com/MicrosoftDocs/azure-docs/issues/38625), so we need to hit the graph API to get their displayName values. This requires the _GroupMember.Read.All_ delegated API permission to be granted to application.
   2. The number of groups is over 200, i.e., group overage. See [how to handle group overage](https://learn.microsoft.com/en-us/azure/active-directory/external-identities/customers/how-to-web-app-role-based-access-control#handle-groups-overage) and [group overages](https://learn.microsoft.com/en-us/security/zero-trust/develop/configure-tokens-group-claims-app-roles#group-overages).

Using this approach, most situations will only require the *User.Read* Microsoft Graph API delegated permission. If it's desired to return display names instead of group IDs (3.i above), the *GroupMember.Read.All* permission is required.

Also update the Azure provider docs:
- general readability improvements
- remove step to modify app manifest for v2.0 as this is not needed (see https://github.com/oauth2-proxy/oauth2-proxy/issues/1666#issuecomment-1333879118 for confusion here)
- add `--azure-graph-group-field` to documentation (added in #1574 but not documented)
- remove `--azure-tenant` from the documentation as this is not necessary because the OIDC discovery URL has what we need

Fixes #1868 #2154

## Motivation and Context

#1574 added support for Azure OAuth 2.0 and querying of groups from the Microsoft Graph API. 

The new groups query that #1574 added cannot be disabled and requires a new API permission, [**GroupMember.Read.All**](https://learn.microsoft.com/en-us/graph/permissions-reference#delegated-permissions-39). This permission requires [administrator consent](https://learn.microsoft.com/en-us/azure/active-directory/develop/permissions-consent-overview?WT.mc_id=Portal-Microsoft_AAD_RegisteredApps#administrator-consent) that would be unnecessary if the groups came from a claim in the ID token. The requirement for administrator consent is particularly painful when using the Azure provider in a multi-tenant scenario, which requires administrators of each tenant to grant consent before their users can use the application.

https://github.com/oauth2-proxy/oauth2-proxy/issues/1666#issuecomment-1297160824 and #1868  show examples of the confusion created by this new requirement.

Additionally, since the groups claim is unused, users who want to limit the groups in the token to only those assigned to the application cannot do so in the current state of the application because all of the user's groups are always returned (see #2154).

## How Has This Been Tested?

1.  [Create an Azure app registration](https://aad.portal.azure.com/#view/Microsoft_AAD_RegisteredApps/CreateApplicationBlade/quickStartType~/null/isMSAApp~/false) with the following settings (do not modify any other settings or grant new API admin permissions):
    - **Name**: Test App without Admin Consent
    - **Supported account types:** Accounts in any organizational directory (Any Azure AD directory - Multitenant)
    - **Redirect URI**: `http://localhost:4180/oauth2/callback`
2. Set client ID and secret environment variables.

    ```bash
    export OAUTH2_PROXY_CLIENT_ID="<client-id>"
    export OAUTH2_PROXY_CLIENT_SECRET="<client-secret>"
    ```

3. Compile and run the proxy with the following settings:

    ```bash
    ./oauth2-proxy \
       --oidc-issuer-url=https://login.microsoftonline.com/organizations/v2.0 \
       --insecure-oidc-skip-issuer-verification \
       --cookie-secret "$(openssl rand -hex 16)" \
       --email-domain "*" \
       --provider azure \
       --redirect-url http://localhost:4180/oauth2/callback
    ```

4. Ensure login at http://localhost:4180 is successful.
5. Restart the proxy with the `--allowed-group "Testing Groups"` and `--azure-graph-group-field "displayName"` flags.

    ```bash
    ./oauth2-proxy \
       --oidc-issuer-url=https://login.microsoftonline.com/organizations/v2.0 \
       --insecure-oidc-skip-issuer-verification \
       --cookie-secret "$(openssl rand -hex 16)" \
       --email-domain "*" \
       --provider azure \
       --redirect-url http://localhost:4180/oauth2/callback \
       --allowed-group "Testing Groups" \
       --azure-graph-group-field "displayName"
    ```

6. Navigate to http://localhost:4180 to initiate the login flow again. This time, a 403 is shown because group validation has been requested, but there are no groups in the returned token (yet).

    image

7. Go to *Token Configuration* in the app and add a groups claim for all groups. See the `OVERVIEW.md` file for specific details.

8. Pick a group ID that your user is a member of and update the command above, removing `--azure-graph-group-field "displayName"`.

    ```bash
    ./oauth2-proxy \
       --oidc-issuer-url=https://login.microsoftonline.com/organizations/v2.0 \
       --insecure-oidc-skip-issuer-verification \
       --cookie-secret "$(openssl rand -hex 16)" \
       --email-domain "*" \
       --provider azure \
       --redirect-url http://localhost:4180/oauth2/callback \
       --allowed-group "39cb3e42-99b7-4404-8747-a515f5865277"
    ```

7. As an administrator, grant the **GroupMember.Read.All** API permission to the app created in step 1 (see the updated doc in this revision for instructions).
10. Navigate to http://localhost:4180 to initiate the login flow again. This time, a 403 error is return, indicating that the user is not in the list of allowed groups.

    image

11. Change the `--allowed-group` flag to a valid Azure group **ID** that I am a member of and restart the proxy.
12. Ensure login at http://localhost:4180 is successful.

## Checklist:

- [x] My change requires a change to the documentation or CHANGELOG.
- [x] I have updated the documentation/CHANGELOG accordingly.
- [x] I have created a feature (non-master) branch for my PR.

### CODE REVIEW

**Comment by rikroe** [comment]
Haven't had the chance to test it out as I cannot build it.

Would this also work if OAuth2 Proxy is started **without** the `--allowed-groups` parameter and we're just passing the allowed groups value directly on the auth url such as `https://some.domain.com/oauth2/auth?allowed_groups=UUID1,UUID2`?

In my understanding this would break because when using `/oauth2/auth` only the existing session is checked and not data from AAD?

If so, I would rather like to have a command line flag that controls this (e.g. `--azure-graph-get-groups`).

File: docs/docs/configuration/overview.md [hunk 1]
```diff
@@ -81,6 +81,7 @@ An example [oauth2-proxy.cfg](https://github.com/oauth2-proxy/oauth2-proxy/blob/
 | `--auth-logging-format` | string | Template for authentication log lines | see [Logging Configuration](#logging-configuration) |
 | `--authenticated-emails-file` | string | authenticate against emails via file (one per line) | |
 | `--azure-tenant` | string | go to a tenant-specific or common (tenant-independent) endpoint. | `"common"` |
+| `--azure_graph_group_field` | string | Microsoft Graph API field to use for Azure group validation against `--allowed-group` ("id" or "displayName") | `"id"` |
```
Comment (rikroe) [inline docs/docs/configuration/overview.md:84]: Shoudn't this be `--azure-graph-group-field`?

Comment (hskiba) [reply docs/docs/configuration/overview.md]: Yes. Thanks

**Comment by hskiba** [comment]
> Would this also work if OAuth2 Proxy is started **without** the `--allowed-groups` parameter and we're just passing the allowed groups value directly on the auth url such as `https://some.domain.com/oauth2/auth?allowed_groups=UUID1,UUID2`?
> 
> In my understanding this would break because when using `/oauth2/auth` only the existing session is checked and not data from AAD?

I believe that is the current behavior in the latest release. That is, the `allowed_groups` query string won't work unless the token happens to be configured to have a groups claim, which it will not by default. This is somewhat related to this comment https://github.com/oauth2-proxy/oauth2-proxy/pull/1574#discussion_r957092288. I didn't know about this query string parameter and its use until you mentioned it, and I'm guessing it wasn't being considered with #1574.

> 
> If so, I would rather like to have a command line flag that controls this (e.g. `--azure-graph-get-groups`).

To be honest, I am not sure I understand why the graph query was added when a groups claim could have been [configured](https://learn.microsoft.com/en-us/azure/active-directory/develop/active-directory-optional-claims#configuring-groups-optional-claims). Could we remove it and update the doc to properly configure the token to have the groups claim if groups validation is desired? Doing that should simplify things while also fixing the functionality mentioned above. For an example from other software, this is how Hashicorp [instructs](https://developer.hashicorp.com/vault/docs/auth/jwt/oidc-providers/azuread#connect-ad-group-with-vault-external-group) (see step 2) configuring Vault to authenticate to Azure if groups validation is desired.

**Comment by gannaramu** [comment]
+1 for this feature. we restirct access at the azure app level and have cases were users have a lot of groups that they are part of and run into large headers issues due to that.

**Comment by rikroe** [comment]
I'm still a little bit at a loss how this all comes together...

My Azure AD app has the `User.Read.All` permission (used it with the OIDC provider before and automatically got the group UUIDs returned).

```sh
docker run --rm -it -p 4180:4180 \
-e OAUTH2_PROXY_CLIENT_ID="$OAUTH2_CLIENT" \
-e OAUTH2_PROXY_CLIENT_SECRET="$OAUTH2_SECRET" \
-e OAUTH2_PROXY_COOKIE_SECRET="$OAUTH2_COOKIE_SECRET" \
quay.io/oauth2-proxy/oauth2-proxy:v7.4.0 \
oauth2-proxy \
--provider=azure \
--redirect-url="http://localhost:4180/oauth2/callback" \
--http-address=0.0.0.0:4180 \
--oidc-issuer-url="https://login.microsoftonline.com/$OAUTH2_TENANT/v2.0" \
--azure-tenant="$OAUTH2_TENANT" \
--email-domain="*" 
```

`oauth2/userinfo` will give something along the lines of
`{"user":"","email":"email@domain.com","groups":["11111111-2222-3333-4444-555555555555"]}`

If I add the following group claims
```
--azure-graph-group-field="displayName" --allowed-group="Internal-Group,11111111-2222-3333-4444-555555555555" 
```

Only the group UUID is checked and `userinfo` returns the same information as above. If I test against only against `Internal-Group`, I get a `403 Forbidden` (because only the group UUIDs are checked).

Oauth2 Proxy also shows the following error message at start up if `--allowed-group` is specified:
```
[2023/01/21 21:58:01] [azure.go:95] WARNING: `groups` scope is not an accepted scope when using Azure OAuth V2 endpoint. Removing it from the scope list
```

This is all using Oauth2 Proxy 7.4.0 from the quay.io docker image.

Oauth2 Proxy however does seem to call `https://graph.microsoft.com/oidc/userinfo` when I'm logging in.

**Comment by hskiba** [comment]
That all makes sense and sounds like the current behavior. Which part specifically are you confused about? It should be switched to just require a groups claim in the token from Azure, I believe. I can attempt to make that switch later today.

Also, I don't think this warning message below is accurate, but I can research and confirm when I attempt to make the change.

> Oauth2 Proxy also shows the following error message at start up if `--allowed-group` is specified:
> 
> ```
> [2023/01/21 21:58:01] [azure.go:95] WARNING: `groups` scope is not an accepted scope when using Azure OAuth V2 endpoint. Removing it from the scope list
> ```

**Comment by rikroe** [comment]
M understanding was that #1574 added the possibility to filter `--allowed-group` via the group's display name instead of UUID. 

Maybe I'm confused as it worked for exactly like that using  OIDC (I need to check the same on 7.3.0).
Because in this case we might not need the second call at all if the data is included anyways?

**Comment by hskiba** [comment]
> M understanding was that #1574 added the possibility to filter `--allowed-group` via the group's display name instead of UUID by setting `--azure-graph-group-field`.

It did a lot more than that I think. It made it so that claims in the token were not used and added the Azure Graph query for groups when a groups claim in the token could have been used (Azure didn't get rid of the ability to add this in v2).

I tried switching the `--azure-graph-group-field` and it was challenging and didn't actually work for me. When I checked what the API was returning, I was getting nulls, which after some research indicated more permission issues. I think this may be why this was defaulted to `id`. Configuring the token to return a groups claims returns names as one would expect.

> Maybe I'm confused as it worked for exactly like that using OIDC (I need to check the same on 7.3.0). Because in this case we might not need the second call at all if the data is included anyways?

Correct. I'll update and it should simplify things further.

**Comment by rikroe** [comment]
Just remembered that there is one valid option to use the additional HTTP call, even if only group UUIDs are returned: being a member in more than 200 groups (c.f. https://github.com/oauth2-proxy/oauth2-proxy/pull/1574#issuecomment-1052451637).

I also did some further digging regarding the group display names (which seem to be only available when using the additional Graph API call).

To see the actual names, the Azure AD application needs additional permissions to `User.Read.All` - see the [List group transitive memberOf documentation](https://learn.microsoft.com/en-us/graph/api/group-list-transitivememberof?view=graph-rest-1.0&tabs=http#permissions).
From those, the `GroupMember.Read.All` least permissive as e.g. `Group.Read.All` also allows access to group calendars (which at least our admins don't like).
I'll try to get one of our admins to give admin consent to `GroupMember.Read.All` and will report back.

So if this works, I think we do have the following options on how to deal with Azure AD logins:
- `v1`: As before, just login without groups
- `v2` **without** `User.Read.All` permission on app: As implemented, but without the graph call. No groups will be returned.
- `v2` **with** `User.Read.All` permission on app: As implemented, but without the graph call. The claim will return the group UUIDs to be used with `--allowed-group` (if a user is member of <200 groups)
- `v2` **with** `User.Read.All` permission on app **and** Graph call: As implemented right now. Group UUIDs (even if >200 groups) will be returned from the Graph API. If `displayName` is requested, it will always be `null`.
- `v2` **with** `User.Read.All` **and** `GroupMember.Read.All` permission on app **and** Graph call: As implemented right now. Group UUIDs (even if >200 groups) and `displayName` will be returned from the Graph API. *Will check if this is actually the case and report back.*

To be honest, I see a reason for all four options of the `v2` call, depending on what is required for the specific use case, what the AAD admins are fine with and how the directory is structured.

As I have no experience in Go at all, I cannot support in implementing anything, but would be happy to support in terms of documentation and trying something out.

**Comment by hskiba** [comment]
In the case of more than 200 groups, I think we should mimic [this approach](https://developer.hashicorp.com/vault/docs/auth/jwt/oidc-providers/azuread#optional-azure-specific-configuration) and document it similarly. I didn't get to updating it this past weekend, but maybe I'll have time this week or this coming weekend.

**Comment by vittoriocanilli** [comment]
On my AKS cluster I use OAuth2-Proxy with Azure provider to authenticate the users and to authorise them to access the Kubernetes namespaces according to the AAD groups, they belong to.

Therefore I agree with @rikroe on this comment in https://github.com/oauth2-proxy/oauth2-proxy/pull/1972#issuecomment-1385377193

> If so, I would rather like to have a command line flag that controls this (e.g. `--azure-graph-get-groups`).

especially if the hypothetical `--azure-graph-get-groups` flag defaults to _true_.

Otherwise if this PR is merged, I feel that it would introduce a breaking change for me and other people, which need the groups: we would need to start using the `--allowed-groups` flag and list there multiple groups.

I agree that a solution is needed for users, who don't need the groups though, but it should not compromise the authentication flow for who need the groups.

**Comment by caizixian** [comment]
I can see #1868 is now automatically closed. This PR once merged will clarify the related documentation a lot.

**Comment by pierluigilenoci** [comment]
Still an issue

**Comment by cameronbriar** [comment]
Leaving a note here in case it's helpful to someone.

Context: My goal was to avoid having the user's groups information in the session info / cookie. In pursuit of implementing something like the `--azure-graph-get-groups` flag described in previous comments above, I discovered there's an undocumented flag, `--azure-graph-group-field` ([1](https://github.com/oauth2-proxy/oauth2-proxy/blob/v7.4.0/pkg/apis/options/legacy_options.go#L486) [2](https://github.com/oauth2-proxy/oauth2-proxy/blob/v7.4.0/docs/docs/configuration/alpha_config.md?plain=1#L167)).

As an experiment, setting this to some unreasonable value (like `notinuse`) partially achieves the goal by substantially reducing the shape/size of the groups information pulled into the session info (because [this](https://github.com/oauth2-proxy/oauth2-proxy/blob/v7.4.0/providers/azure.go#L401) results in an empty slice) -- resulting in a useful hack that allows me to continue using cookies for storage even in situations where the group membership is enormous (which is problematic in some scenarios).

Note: Having a flag like `--azure-graph-get-groups` is still the preferred solution.

**Comment by hskiba** [comment]
I added the discussed `--azure-graph-get-groups` flag for deployments that do not use the `--allowed-groups` flag but still want to query group information and store it in the oauth2-proxy session. I also tested that only a delegated _GroupMember.Read.All_ permission is required when querying group membership for a user.

**Comment by hskiba** [comment]
Going to refactor the Azure provider a bit:

1. Use the [Microsoft Graph SDK for Go](https://github.com/microsoftgraph/msgraph-sdk-go).
2. Use the `groups` claim if it's present.
3. Only query the Microsoft Graph API in the following cases:
   1. `--azure-graph-group-field` is `displayName`. It's only possible to return group IDs in a groups claim (see https://github.com/MicrosoftDocs/azure-docs/issues/38625), so we need to hit the graph API to get their displayName values.
   2. The number of groups is over 200, i.e., group overage. See [how to handle group overage](https://learn.microsoft.com/en-us/azure/active-directory/external-identities/customers/how-to-web-app-role-based-access-control#handle-groups-overage) and [group overages](https://learn.microsoft.com/en-us/security/zero-trust/develop/configure-tokens-group-claims-app-roles#group-overages).

Using this approach, most situations will only require the *User.Read* Microsoft Graph API delegated permission. If it's desired to return display names instead of group IDs (3.i above), the *GroupMember.Read.All* permission is required.

Making these changes will also address #2154.

**Comment by pierluigilenoci** [comment]
@JoelSpeed could you please take a look?

**Comment by pierluigilenoci** [comment]
@hskiba, according to CodeClimate, your code needs a bit of reshaping.

**Comment by hskiba** [comment]
> @hskiba, according to CodeClimate, your code needs a bit of reshaping.

Yes, still in draft mode. I was trying to prevent it from getting marked stale as it's still an issue.

**Comment by hskiba** [comment]
Haven't had a chance to get back to this. Should have some time soon. It's close.

**Comment by jjlakis** [comment]
Hey @hskiba How is to going? I am about have a deeper look into refactoring Azure Provider class, it would be nice to have this change ready beforehand :man_dancing:

**Comment by hskiba** [comment]
> Hey @hskiba How is to going? I am about have a deeper look into refactoring Azure Provider class, it would be nice to have this change ready beforehand 🕺

No updates. Hope to have some time this weekend.

**Comment by hskiba** [comment]
Going to abandon this in favor of #2390. #2390 suggests keeping this around as a legacy provider, but I made similar changes as #2390 that would effectively remove that legacy support. With the landing of #2390, I wouldn't recommend keeping the Azure provider around as "legacy" as the upgrade to use the OIDC compliant version is simple.

**Pull request closed without merging**